### PR TITLE
Feature/205 test runner

### PR DESCRIPTION
**Pull request by awinia-github**

### CODE REVIEW

File: src/ATE_sammy/ate_sammy/templates/test_runner_main_template.jinja2 [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        assert len(testers) == 1, "The testertype {{hardware_definition["tester"]}} is not available or installed on this machine."
+        self.tester = testers[0]
+
+        {% for actuator in hardware_definition['Actuator'][hardware_definition['base']] %}
+        from ate_test_app.actuators.{{actuator}}.{{actuator}} import {{actuator}}Proxy
+        self.{{actuator}}_instance = {{actuator}}Proxy()
+        self.{{actuator}}_instance.set_mqtt_client(self.harness)
+        {% endfor %}
+
+        # TODO: do we need to suppot extenal instrument
```
Comment (awinia-github) [inline src/ATE_sammy/ate_sammy/templates/test_runner_main_template.jinja2:55]: Would be nice if you add some issue number or discussion number.

File: src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py [hunk 2]
```diff
@@ -1,16 +1,521 @@
...
+    row: int
+    col: int
+
+
+class TabInterface(ABC):
+    def __init__(self, parent: 'TestRunner'):
+        self.parent = parent
+
+    @abstractclassmethod
+    def setup_view(self):
```
Comment (awinia-github) [inline src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py:108]: return type

File: src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py [hunk 3]
```diff
@@ -1,16 +1,521 @@
...
+
+
+@dataclass
+class CellInfo:
+    row: int
+    col: int
+
+
+class TabInterface(ABC):
+    def __init__(self, parent: 'TestRunner'):
```
Comment (awinia-github) [inline src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py:104]: return type

File: src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py [hunk 4]
```diff
@@ -1,16 +1,521 @@
...
+        self.parent = parent
+
+    @abstractclassmethod
+    def setup_view(self):
+        pass
+
+    @abstractclassmethod
+    def setup_callbacks(self):
```
Comment (awinia-github) [inline src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py:112]: return type

File: src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py [hunk 5]
```diff
@@ -1,16 +1,521 @@
...
+        pass
+
+
+class RunTab(TabInterface):
+    def __init__(self, parent: 'TestRunner'):
```
Comment (awinia-github) [inline src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py:117]: return type

File: src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py [hunk 6]
```diff
@@ -1,16 +1,521 @@
...
+        super().__init__(parent)
+        self.current_cell_info: Optional[CellInfo] = None
+        self.input_table_data = InputTable()
+        self.output_table_data = []
+
+    def setup_view(self):
```
Comment (awinia-github) [inline src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py:123]: return type

File: src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py [hunk 7]
```diff
@@ -1,16 +1,521 @@
...
+        self.parent.inputParameter.horizontalHeader().setSectionResizeMode(QtWidgets.QHeaderView.Stretch)
+        self.parent.outputParameter.horizontalHeader().setSectionResizeMode(QtWidgets.QHeaderView.Stretch)
+
+        self.parent.inputParameter.setSelectionMode(QtWidgets.QAbstractItemView.SingleSelection)
+        self.parent.outputParameter.setSelectionMode(QtWidgets.QAbstractItemView.SingleSelection)
+
+        self.parent.inputParameter.selectionModel().selectionChanged.connect(lambda selected, deselected: self._update_cell(selected, deselected, self.parent.inputParameter))
+        self.parent.outputParameter.selectionModel().selectionChanged.connect(lambda selected, deselected: self._update_cell(selected, deselected, self.parent.outputParameter))
+
+    def setup_callbacks(self):
```
Comment (awinia-github) [inline src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py:146]: return type

File: src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py [hunk 8]
```diff
@@ -1,16 +1,521 @@
...
+        # TODO: the implementation of this part shall be done in #239
+        self.parent.start.clicked.connect(self._start_execution)
+        ''''''
+
+    def _start_execution(self):
```
Comment (awinia-github) [inline src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py:151]: return type

File: src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py [hunk 9]
```diff
@@ -1,16 +1,521 @@
...
+        # TODO: the code generation may be the same as generating a test program but with only one test
+        # TODO: emit signal to spyder
+        self.parent.feedback.setText('not implemented yet')
+
+    def _fill_output_parameter_table(self):
```
Comment (awinia-github) [inline src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py:156]: return type

File: src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py [hunk 10]
```diff
@@ -1,16 +1,521 @@
...
+from ate_common.parameter import InputColumnKey, OutputColumnKey
+
+from PyQt5 import QtCore
+from PyQt5 import QtGui
+from PyQt5 import QtWidgets
+
+
+@dataclass
+class InputRow:
+    __slot__ = ['name', 'shmoo', 'min', 'call', 'step', 'max', 'exponent', 'unit']
```
[the remark this answers is not in the document]
Comment (awinia-github) [reply src/ATE_spyder/ate_spyder/widgets/actions_on/tests/TestRunner.py:22]: as discussed with ar and sk keep it like it is

**Review by nerohmot — APPROVED** [review]

**Pull request merged**